database2/ffboltdb/interface_test.go

```diff
@@ -0,0 +1,2266 @@
+			badBlockHash)
+		_, err = tx.FetchBlockHeader(badBlockHash)
+		if !checkDbError(tc.t, testName, err, wantErrCode) {
+			return false
+		}
+
+		// Ensure fetching a block region in a block that doesn't exist
+		// return the expected error.
+		testName = fmt.Sprintf("FetchBlockRegion(%s) invalid hash",
+			badBlockHash)
+		wantErrCode = database.ErrBlockNotFound
+		region.Hash = badBlockHash
+		region.Offset = ^uint32(0)
+		_, err = tx.FetchBlockRegion(&region)
+		if !checkDbError(tc.t, testName, err, wantErrCode) {
+			return false
+		}
+
+		// Ensure fetching a block region that is out of bounds returns
+		// the expected error.
+		testName = fmt.Sprintf("FetchBlockRegion(%s) invalid region",
+			blockHash)
+		wantErrCode = database.ErrBlockRegionInvalid
+		region.Hash = blockHash
+		region.Offset = ^uint32(0)
+		_, err = tx.FetchBlockRegion(&region)
+		if !checkDbError(tc.t, testName, err, wantErrCode) {
+			return false
+		}
+	}
+
+	// -----------------
+	// Bulk Block IO API
+	// -----------------
+
+	// Ensure the bulk block data fetched from the database matches the
+	// expected bytes.
+	blockData, err := tx.FetchBlocks(allBlockHashes)
+	if err != nil {
+		tc.t.Errorf("FetchBlocks: unexpected error: %v", err)
+		return false
+	}
+	if len(blockData) != len(allBlockBytes) {
+		tc.t.Errorf("FetchBlocks: unexpected number of results - got "+
+			"%d, want %d", len(blockData), len(allBlockBytes))
+		return false
+	}
+	for i := 0; i < len(blockData); i++ {
+		blockHash := allBlockHashes[i]
+		wantBlockBytes := allBlockBytes[i]
+		gotBlockBytes := blockData[i]
+		if !bytes.Equal(gotBlockBytes, wantBlockBytes) {
+			tc.t.Errorf("FetchBlocks(%s): bytes mismatch: got %x, "+
+				"want %x", blockHash, gotBlockBytes,
+				wantBlockBytes)
+			return false
+		}
+	}
+
+	// Ensure the bulk block headers fetched from the database match the
+	// expected bytes.
+	blockHeaderData, err := tx.FetchBlockHeaders(allBlockHashes)
+	if err != nil {
+		tc.t.Errorf("FetchBlockHeaders: unexpected error: %v", err)
+		return false
+	}
+	if len(blockHeaderData) != len(allBlockBytes) {
+		tc.t.Errorf("FetchBlockHeaders: unexpected number of results "+
+			"- got %d, want %d", len(blockHeaderData),
+			len(allBlockBytes))
+		return false
+	}
+	for i := 0; i < len(blockHeaderData); i++ {
+		blockHash := allBlockHashes[i]
+		wantHeaderBytes := allBlockBytes[i][0:wire.MaxBlockHeaderPayload]
+		gotHeaderBytes := blockHeaderData[i]
+		if !bytes.Equal(gotHeaderBytes, wantHeaderBytes) {
+			tc.t.Errorf("FetchBlockHeaders(%s): bytes mismatch: "+
+				"got %x, want %x", blockHash, gotHeaderBytes,
+				wantHeaderBytes)
+			return false
+		}
+	}
+
+	// Ensure the first transaction of every block fetched in bulk block
+	// regions from the database matches the expected bytes.
+	allRegionBytes, err := tx.FetchBlockRegions(allBlockRegions)
+	if err != nil {
+		tc.t.Errorf("FetchBlockRegions: unexpected error: %v", err)
+
```

Comment:
Yep missing a return there.  It catches the mismatch just below, but better to be consistent.